core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -248,34 +248,25 @@ sealed abstract class IO[E, A] { self =>
   final def ensuring(finalizer: Infallible[Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
-  /**
-   * Executes the release action only if there was an error.
+  /**	
+   * Executes the release action only if there was an error.	
    */
   final def bracketOnError[B](release: A => Infallible[Unit])(use: A => IO[E, B]): IO[E, B] =
-    bracket0(
-      (r: ExitResult[E, B], a: A) =>
-        r match {
-          case ExitResult.Failed(_)     => release(a)
-          case ExitResult.Terminated(_) => release(a)
-          case _                        => IO.unit
-      }
-    )(use)
+    IO.bracket0(this)((a: A, _: Option[Either[E, B]]) => release(a))(use)
 
   /**
    * Runs one of the specified cleanup actions if this action errors, providing the
    * error to the cleanup action. The cleanup action will not be interrupted.
    * Cleanup actions for handled and unhandled errors can be provided separately.
    */
-  final def onError(cleanupT: Throwable => Infallible[Unit])(cleanupE: E => Infallible[Unit]): IO[E, A] =
-    IO.unit[E]
-      .bracket0(
-        (r: ExitResult[E, A], a: Unit) =>
-          r match {
-            case ExitResult.Failed(e)     => cleanupE(e)
-            case ExitResult.Terminated(t) => cleanupT(t)
-            case _                        => IO.unit
-        }
-      )(_ => self)
+  final def onError(cleanupE: E => Infallible[Unit]): IO[E, A] =
+    IO.bracket0(IO.unit[E])(
```

Comment:
I would do `Option[E]` here, and if the provided option is `None`, then call release using `None`.